ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/PaginatingAsyncStream.scala

```diff
@@ -51,33 +50,30 @@ object PaginatingAsyncStream {
     * a single [[Source]], passing the last seen event's ledger [[Offset]] and node index to the
     * next iteration query, so it can continue reading events from this point.
     *
-    * This is to implement pagination based on ledger offset and event node index.
+    * This is to implement pagination based on generic offset.
     * The main purpose of the pagination is to break down large queries
     * into smaller batches. The reason for this is that we are currently using
     * simple blocking JDBC APIs and a long-running stream would end up
     * occupying a thread in the DB pool, severely limiting the ability
     * of keeping multiple, concurrent, long-running streams while serving
     * lookup calls.
     *
-    * @param initialOffset initial ledger [[Offset]]
-    * @param extractOffsetAndNodeIndex function that extracts [[Offset]] and node index from the result entry of type [[T]]
-    * @param query a function that takes [[Offset]] and optional node index to start pagination from
+    * @param startFromOffset initial offset
+    * @param nextPageOffset function that returns next page offset or [[None]] when no more pagination is required
+    * @param query a function that fetches results starting from provided offset
+    * @tparam Off the type of the offset
     * @tparam T the type of the items returned in each call
     */
-  def streamFrom[T](initialOffset: Offset, extractOffsetAndNodeIndex: T => (Offset, Int))(
-      query: (Offset, Option[Int]) => Future[Vector[T]]
+  def streamFrom[Off, T](startFromOffset: Off, nextPageOffset: Vector[T] => Option[Off])(
```

Comment:
I think you could keep the logic around `lastOption` within this class and make `nextPageOffset: T => Option[Off]`. I expect we always will want to use the last element anyway.

Reply:
I think it should be `T => Off` but I would not call it `nextPageOffset` in this case... it would be extract offset, or just `f` -- the type should be enough

Reply:
in any case, I like I agree with you @gerolf-da it always has to be the last element. I think the new signature looks better... I would also consider removing the scaladoc for this function :)... I think it is clear from the function signature what this function does.